pkg/vsphere/vm/vm.go

```diff
@@ -512,23 +525,70 @@ func (vm *VirtualMachine) needsFix(err error) bool {
 	case *types.InvalidState:
 		return true
 	default:
+		if vm.IsInvalidState(ctx) {
+			log.Debugf("vm %s is invalid", vm.Reference())
+			return true
+		}
 		log.Debugf("Do not fix non invalid state error")
 		return false
 	}
 }
 
+func (vm *VirtualMachine) IsInvalidState(ctx context.Context) bool {
+	var o mo.VirtualMachine
+	if err := vm.VirtualMachine.Properties(ctx, vm.Reference(), []string{connectionStatus}, &o); err != nil {
+		log.Debugf("Failed to get vm properties: %s", err)
+		return false
+	}
+	if o.Summary.Runtime.ConnectionState == types.VirtualMachineConnectionStateInvalid {
+		return true
+	}
+	return false
+}
+
 // WaitForResult is designed to handle VM invalid state error for any VM operations.
 // It will call tasks.WaitForResult to retry if there is task in progress error.
 func (vm *VirtualMachine) WaitForResult(ctx context.Context, f func(context.Context) (tasks.Task, error)) (*types.TaskInfo, error) {
 	info, err := tasks.WaitForResult(ctx, f)
-	if err == nil || !vm.needsFix(err) {
+	if err == nil || !vm.needsFix(ctx, err) {
 		return info, err
 	}
 	log.Debugf("Try to fix task failure %s", err)
-	if nerr := vm.FixInvalidState(ctx); nerr != nil {
+	if nerr := vm.fixVM(ctx); nerr != nil {
 		log.Errorf("Failed to fix task failure: %s", nerr)
 		return info, err
 	}
 	log.Debugf("Fixed")
 	return tasks.WaitForResult(ctx, f)
 }
+
+func (vm *VirtualMachine) Properties(ctx context.Context, r types.ManagedObjectReference, ps []string, o *mo.VirtualMachine) error {
+	log.Debugf("get vm properties %s of vm %s", ps, r)
+	contains := false
+	for i := range ps {
+		if ps[i] == summary {
+			contains = true
+			break
+		}
+	}
+	var newps []string
```

Comment:
It doesn't matter to query duplicated properties, but as we're trying to make minimum properties query each time, it will help to reduce the array length.
And I think this array length is really short based on our current coding practice, I assume it doesn't hurt to loop that array.